cache/disk/disk.go

```diff
@@ -377,18 +379,27 @@ func (c *Cache) loadExistingFiles() error {
 	log.Println("Building LRU index.")
 	for _, f := range files {
 		relPath := f.name[len(c.dir)+1:]
-		var lookupKey string
 
-		sizeOnDisk := f.info.Size()
-		size := sizeOnDisk
+		legacy := strings.HasSuffix(f.name, ".v1")
 
 		fields := strings.Split(relPath, "/")
+
 		hash := fields[len(fields)-1]
+		if legacy {
+			hash = strings.TrimSuffix(hash, ".v1")
+		}
+
+		sizeOnDisk := f.info.Size()
+		size := sizeOnDisk
+
+		var lookupKey string
 
 		if strings.HasPrefix(relPath, "cas.v2/") {
-			size, err = casblob.GetLogicalSize(f.name)
-			if err != nil {
-				return err
+			if !legacy {
+				size, err = casblob.GetLogicalSize(f.name)
```

Comment:
In my largest bazel-remote instance that's running something close to this branch, restarting is fast enough. It's still growing though, so I can re-check once it has filled the space.

Reply:
The CAS in one of our cache instances is 1500 GB and contains 8 000 000 files.

I’m emulating restarting the cache by reading 8 bytes from them. Running “xargs head -c 8 > /dev/null” over them took 20 minutes. That is too long for us.

Re-running again with warm file system caches took 45 seconds.

(A bazel-remote instance was using the CAS at the same time)

Some of our caches are even larger. All of them are on SSD, but other people might have rotating disks where these 8 byte random-accesses takes even more time.

Reply:
I think a 45 second restart for a large cache is reasonable, though not ideal.

A 20 minute cold start is pretty bad though. There is one other potential startup optimization that could help- loading file details concurrently (I know that @bdittmer has a local patch for this, which helps even at small scale). I wonder if you could try using `xargs -P 72` in your benchmark (and maybe a couple of lower values too)?

(I don't think we should bother considering HDD-speeds, it's the wrong tool for the job and I would rather just recommend consumer SSD-like speed storage as a minimum.)

Reply:
45 second restart time is longer than I would like in production. Because the longer time the cache is down, while restarting to upgrade version or change configuration, the higher risk for failing builds-without-the-bytes assumption about cache always available.

A first try with -P 72 took 4 minutes, but when re-trying with -P 72 more times, it took 1m45s, despite clearing filesystem cache with "sync; echo 3 > /proc/sys/vm/drop_caches" in between. Due to the difference, I don't fully trust the numbers. I wonder if there are some other caching layer involved, or if the first try coincided with peak load from the bazel-remote instance also running towards the disk. The fastest time I got was with -P 64, which was 1m 17s. If you are interested, I can list values for all -P values I tried.

But it come to my mind that there is one more large issue: **Reading headers from files during startup, would touch file system access time, and that affects the ordering in LRU.**

I believe all these issues can be avoided by simply storing the logical size embedded in the file name, like: cas/ef95d34495b36c0a6108c4e48679ebebfb1cee01364e3dceead1f95ae4276d2b_**450**.zst, if the logical size is 450 bytes and compressed with zstandard.

Or even better, I would prefer having both the logical size and the on-disk-size embedded in the file name. Since if there is a guaranteed one-to-one association between disk.lruItem entries and file names, then we can consider releasing the mutex during some syscalls in disk.go, as discussed in #323. Any changes to the mutexes could be considered in later separate work, but embedding both sizes, as preparation, would be easier in this PR.

Example zstandard compressed with logical size 450 bytes and on-disk-size 290 bytes:

`cas/ef95d34495b36c0a6108c4e48679ebebfb1cee01364e3dceead1f95ae4276d2b_450_290.zst`

Example non-compressed with both logical and on-disk size as 450 bytes:

`cas/ef95d34495b36c0a6108c4e48679ebebfb1cee01364e3dceead1f95ae4276d2b_450`

Reply:
Those numbers look bad enough to give up on reading all the headers during loading. I will start refactoring.

I don't think the combination of logical size and on-disk size would be unique enough for the early mutex release idea- that would probably require storing a pseudo-random string or number in each lruItem.

Aside: using atimes for the initial LRU order is only approximate, because `Contains` updates the LRU but for efficiency reasons does not touch the filesystem (builds-without-the-bytes makes this worse). Reading data from file headers didn't make things much worse because the files were sorted by atime in one pass, then imported in that order in a second pass (but it added bias to the filesystem atime order for compressed CAS items). If the cache was running long enough then the atime order would become more similar to the LRU index order.

In the future, I might consider writing out the index order to disk prior to clean shutdown (and maybe also periodically) and then only use atimes as a fallback.

Reply:
> I don't think the combination of logical size and on-disk size would be unique enough for the early mutex release

I believe it would be enough, since then it would not matter how concurrent PUT requests of same key are interleaved. Because all their LRU entry commits would be compatible with any of their files, and it would not matter if we end up with LRU commit from one request and file from another. PUT requests interleaved with evictions of same key, is more complex, but I have hope that is also possible without holding mutexes during file system syscalls.

> Reading data from file headers didn't make things much worse because the files were sorted by atime in one pass, then imported in that order in a second pass 

Our CAS file systems are mounted with the relatime option, so each access time is not updated more than once a day.  That is normally not an issue, since files remains in our caches much longer than a day. But I suspect that same order in first and second pass would not have been sufficient with relatime.

Thanks for the refactoring!